File: 2024/day-1/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

// Read input file and parse numbers into two slices
func readInput(filePath string) ([]int, []int, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, nil, err
	}
	defer file.Close()

	var left, right []int
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		parts := strings.Fields(scanner.Text()) // Split by spaces
		if len(parts) != 2 {
			return nil, nil, fmt.Errorf("invalid line format: %s", scanner.Text())
		}

		a, err1 := strconv.Atoi(parts[0])
		b, err2 := strconv.Atoi(parts[1])
		if err1 != nil || err2 != nil {
			return nil, nil, fmt.Errorf("invalid numbers in line: %s", scanner.Text())
		}

		left = append(left, a)
		right = append(right, b)
	}

	if err := scanner.Err(); err != nil {
		return nil, nil, err
	}

	return left, right, nil
}

// Part 1: Calculate the sum of absolute differences
func minDifferenceSum(left, right []int) int {
	sort.Ints(left)
	sort.Ints(right)

	sum := 0
	for i := 0; i < len(left); i++ {
		diff := left[i] - right[i]
		if diff < 0 {
			diff = -diff // Absolute value
		}
		sum += diff
	}

	return sum
}

// Part 2: Calculate similarity score
func similarityScore(left, right []int) int {
	// Create a frequency map for right list
	countRight := make(map[int]int)
	for _, num := range right {
		countRight[num]++
	}

	// Calculate similarity score
	score := 0
	for _, num := range left {
		score += num * countRight[num]
	}

	return score
}

func main() {
	filePath := "input.txt"

	// Read and parse input
	leftList, rightList, err := readInput(filePath)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Solve Part 1
	result1 := minDifferenceSum(leftList, rightList)
	fmt.Println(result1)

	// Solve Part 2
	result2 := similarityScore(leftList, rightList)
	fmt.Println(result2)
}

File: 2024/day-1/solution.py
from collections import Counter

def min_difference_sum(left, right):
    left.sort()
    right.sort()
    return sum(abs(a - b) for a, b in zip(left, right))

def read_input(file_path):
    left, right = [], []
    with open(file_path, 'r') as file:
        for line in file:
            a, b = map(int, line.split())
            left.append(a)
            right.append(b)
    return left, right  # Move return statement outside 'with'

    
def similarity_score(left, right):
    count_right = Counter(right)  # O(N)
    return sum(num * count_right.get(num, 0) for num in left)  # O(N)

file_path = "input.txt"
left_list, right_list = read_input(file_path)

result_1 = min_difference_sum(left_list, right_list) # Part 1
result_2 = similarity_score(left_list, right_list) # Part 2

print(result_1)
print(result_2)